# 20min SFL Fantasy

Du willst gerade auch das optimale Team finden, um beim 20min Fantasy zu gewinnen? Ich habe dir hier ein paar Stats dazu.

In [106]:
import requests
import json
import pandas as pd

url = "https://fantasy.20min.ch/json/fantasy/stats_centre_table.json"

try:
    response = requests.get(url)
    response.raise_for_status()  # Raise an exception for bad status codes
    data = response.json()
    df = pd.DataFrame(data)
    # You can now work with the 'data' variable which contains the JSON content
    # print(data) # Uncomment this line to see the data
except requests.exceptions.RequestException as e:
    print(f"Error fetching data: {e}")
except json.JSONDecodeError as e:
    print(f"Error decoding JSON: {e}")

print(f"Anzahl Spieler: {len(data)}")
print(df.columns)

Anzahl Spieler: 312
Index(['playerId', 'squadId', 'firstName', 'lastName', 'displayName',
       'position', 'price', 'gameWeekPoints', 'averagePoints', 'last3Average',
       'bonusPoints', 'totalPoints', 'percentSelected', 'form', 'goals',
       'assists', 'minutesPlayed', 'shotsOnTarget', 'chancesCreated',
       'tackles', 'cleanSheet', 'saves', 'goalsConceded', 'yellowCards',
       'redCards', 'ownGoals', 'penaltyMisses', 'penaltySaves', 'bonusPpm',
       'dribbles', 'crosses', 'offsides', 'passCompletionRate',
       'interceptions', 'blocks', 'foulsWon', 'foulsMade', 'goalsOutsideArea',
       'errorsLeadingToGoal'],
      dtype='object')


Schaeun wir zuerst auf die totalen Punkte:

In [155]:
num = 11
print(f"Top {num} by Points")
res = df.sort_values(by=['totalPoints', 'price'], ascending=[False, True])[['displayName', 'position', 'totalPoints', 'price']].head(num)
print(res['position'].value_counts())
display(res)

Top 11 by Points
position
MID    5
GK     4
FWD    1
DEF    1
Name: count, dtype: int64


displayName position  totalPoints  price
5       X. Shaqiri      MID          265   11.0
7    M. Stevanovic      MID          224   10.0
2          M. Hitz       GK          183    7.5
14      Y. Brecher       GK          181    7.5
176      B. Traoré      FWD          170    9.0
90       D. Schmid      DEF          165    8.0
103      J. Hammel       GK          164    7.0
10       S. Kapino       GK          160    7.0
174     A. Sanches      MID          158    8.5
178    D. Rrudhani      MID          150    7.5
147   M. Di Giusto      MID          149    8.5

In [162]:
# Calculate the ratios
df['averagePoints_per_price'] = df['averagePoints'] / df['price']
df['totalPoints_per_price'] = df['totalPoints'] / df['price']

num = 20

# best avg
print(f"Top {num} by Average Points per Price:")
avg_df = df[(df['position']!= 'GK') & (df['minutesPlayed'] > 900)]

res = avg_df.sort_values(by=['averagePoints_per_price', 'price'], ascending=[False, True])[['displayName', 'position', 'totalPoints', 'averagePoints', 'price', 'averagePoints_per_price', 'minutesPlayed']].head(num)
print(res['position'].value_counts())
display(res)

# worst avg
print(f"\nWorst {num} by Average Points per Price:")
res = df.sort_values(by=['averagePoints_per_price', 'price'], ascending=[True, False])[['displayName', 'position', 'totalPoints', 'averagePoints','price', 'averagePoints_per_price']].head(num)
print(res['position'].value_counts())
display(res)

# best total
print(f"\nTop {num}  by Total Points per Price:")
res = df.sort_values(by=['totalPoints_per_price', 'price'], ascending=[False, True])[['displayName', 'position','totalPoints', 'averagePoints', 'price', 'totalPoints_per_price']].head(num)
print(res['position'].value_counts())
display(res)

# worst total
print(f"\nWorst {num}  by Total Points per Price:")
res = df.sort_values(by=['totalPoints_per_price', 'price'],  ascending=[True, False])[['displayName', 'position','totalPoints', 'averagePoints', 'price', 'totalPoints_per_price']].head(num)
print(res['position'].value_counts())
display(res)


Top 20 by Average Points per Price:
position
DEF    10
MID     9
FWD     1
Name: count, dtype: int64


displayName position  totalPoints  averagePoints  price  \
5       X. Shaqiri      MID          265            7.8   11.0   
142       P. Otele      MID           95            5.3    8.0   
174     A. Sanches      MID          158            5.6    8.5   
28      B. Kololli      MID           95            4.5    7.0   
178    D. Rrudhani      MID          150            4.8    7.5   
12     S. Rouiller      DEF          112            4.0    6.5   
72    C. Fassnacht      MID          104            5.8    9.5   
7    M. Stevanovic      MID          224            5.9   10.0   
90       D. Schmid      DEF          165            4.7    8.0   
271       L. Lüthi      DEF           92            3.5    6.0   
4        L. Benito      DEF           86            3.2    5.5   
258       C. Tsawa      MID           47            2.9    5.0   
16     N. Dussenne      DEF          143            4.3    7.5   
207         K. Sow      DEF           83            2.8    5.0   
151     A. Barisic      DEF           99            3.3    6.0   
52      K. Hajrizi      DEF           36            2.4    4.5   
292    B. Freimann      DEF           47            2.4    4.5   
80        M. Poaty      DEF          117            3.2    6.0   
1         S. Zuber      MID           94            4.7    9.0   
176      B. Traoré      FWD          170            4.7    9.0   

     averagePoints_per_price  minutesPlayed  
5                   0.709091           2618  
142                 0.662500           1020  
174                 0.658824           2371  
28                  0.642857           1208  
178                 0.640000           2189  
12                  0.615385           2419  
72                  0.610526           1440  
7                   0.590000           3179  
90                  0.587500           2922  
271                 0.583333           2340  
4                   0.581818           2246  
258                 0.580000           1000  
16                  0.573333           2937  
207                 0.560000           2320  
151                 0.550000           2424  
52                  0.533333           1272  
292                 0.533333           1409  
80                  0.533333           2956  
1                   0.522222           1767  
176                 0.522222           2411


Worst 20 by Average Points per Price:
position
FWD    8
MID    6
DEF    5
GK     1
Name: count, dtype: int64


displayName position  totalPoints  averagePoints  price  \
17     E. Alioski      DEF            0            0.0    7.0   
183   A. Kendouci      MID            0            0.0    7.0   
49        S. Mráz      FWD            0            0.0    6.5   
137  N. Reichmuth      MID            0            0.0    6.5   
141       B. Okoh      DEF            0            0.0    6.5   
188   R. Nivokazi      FWD            0            0.0    6.5   
274     M. Gutbub      FWD            0            0.0    6.5   
22     L. Bertone      MID            0            0.0    6.0   
33    G. Wüthrich      DEF            0            0.0    6.0   
84    A. Racioppi       GK            0            0.0    6.0   
193   E. Rastoder      FWD            0            0.0    6.0   
291   E. Meichtry      MID            0            0.0    6.0   
51       C. Ibayi      FWD            0            0.0    5.5   
67   D. Jovanovic      FWD            0            0.0    5.5   
78    V. Janjicic      MID            0            0.0    5.5   
97    G. Montolio      DEF            0            0.0    5.5   
104       J. Roth      MID            0            0.0    5.5   
120    L. Stewart      FWD            0            0.0    5.5   
182   A. Hunziker      FWD            0            0.0    5.5   
231       T. Gaal      DEF            0            0.0    5.5   

     averagePoints_per_price  
17                       0.0  
183                      0.0  
49                       0.0  
137                      0.0  
141                      0.0  
188                      0.0  
274                      0.0  
22                       0.0  
33                       0.0  
84                       0.0  
193                      0.0  
291                      0.0  
51                       0.0  
67                       0.0  
78                       0.0  
97                       0.0  
104                      0.0  
120                      0.0  
182                      0.0  
231                      0.0


Top 20  by Total Points per Price:
position
GK     9
DEF    6
MID    4
FWD    1
Name: count, dtype: int64


displayName position  totalPoints  averagePoints  price  \
2          M. Hitz       GK          183            5.1    7.5   
14      Y. Brecher       GK          181            4.8    7.5   
5       X. Shaqiri      MID          265            7.8   11.0   
103      J. Hammel       GK          164            4.4    7.0   
10       S. Kapino       GK          160            4.8    7.0   
7    M. Stevanovic      MID          224            5.9   10.0   
73       K. Letica       GK          142            4.2    6.5   
90       D. Schmid      DEF          165            4.7    8.0   
50         L. Zigi       GK          132            3.5    6.5   
197      P. Loretz       GK          132            3.5    6.5   
178    D. Rrudhani      MID          150            4.8    7.5   
80        M. Poaty      DEF          117            3.2    6.0   
138       A. Saipi       GK          117            3.4    6.0   
16     N. Dussenne      DEF          143            4.3    7.5   
60         P. Dorn      DEF          133            3.6    7.0   
176      B. Traoré      FWD          170            4.7    9.0   
174     A. Sanches      MID          158            5.6    8.5   
129      T. Fayulu       GK          110            3.1    6.0   
9      N. Lavanchy      DEF          100            2.8    5.5   
101       M. Gómez      DEF          109            2.9    6.0   

     totalPoints_per_price  
2                24.400000  
14               24.133333  
5                24.090909  
103              23.428571  
10               22.857143  
7                22.400000  
73               21.846154  
90               20.625000  
50               20.307692  
197              20.307692  
178              20.000000  
80               19.500000  
138              19.500000  
16               19.066667  
60               19.000000  
176              18.888889  
174              18.588235  
129              18.333333  
9                18.181818  
101              18.166667


Worst 20  by Total Points per Price:
position
FWD    8
MID    6
DEF    5
GK     1
Name: count, dtype: int64


displayName position  totalPoints  averagePoints  price  \
17     E. Alioski      DEF            0            0.0    7.0   
183   A. Kendouci      MID            0            0.0    7.0   
49        S. Mráz      FWD            0            0.0    6.5   
137  N. Reichmuth      MID            0            0.0    6.5   
141       B. Okoh      DEF            0            0.0    6.5   
188   R. Nivokazi      FWD            0            0.0    6.5   
274     M. Gutbub      FWD            0            0.0    6.5   
22     L. Bertone      MID            0            0.0    6.0   
33    G. Wüthrich      DEF            0            0.0    6.0   
84    A. Racioppi       GK            0            0.0    6.0   
193   E. Rastoder      FWD            0            0.0    6.0   
291   E. Meichtry      MID            0            0.0    6.0   
51       C. Ibayi      FWD            0            0.0    5.5   
67   D. Jovanovic      FWD            0            0.0    5.5   
78    V. Janjicic      MID            0            0.0    5.5   
97    G. Montolio      DEF            0            0.0    5.5   
104       J. Roth      MID            0            0.0    5.5   
120    L. Stewart      FWD            0            0.0    5.5   
182   A. Hunziker      FWD            0            0.0    5.5   
231       T. Gaal      DEF            0            0.0    5.5   

     totalPoints_per_price  
17                     0.0  
183                    0.0  
49                     0.0  
137                    0.0  
141                    0.0  
188                    0.0  
274                    0.0  
22                     0.0  
33                     0.0  
84                     0.0  
193                    0.0  
291                    0.0  
51                     0.0  
67                     0.0  
78                     0.0  
97                     0.0  
104                    0.0  
120                    0.0  
182                    0.0  
231                    0.0

In [152]:
df.to_csv("data.csv")

Alter chart nehmen! und dann einfach jitter einbauen

In [167]:
import pandas as pd
import altair as alt
import numpy as np

# Anzahl der Top-Spieler pro Position
num = 10

# Jitter hinzufügen
df['price_jitter'] = df['price'] + np.random.uniform(-0.1, 0.1, size=len(df))
df['totalPoints_per_price'] = df['totalPoints'] / df['price']

# Top-N Spieler pro Position (z. B. 10)
top_per_position = (
    df.sort_values(by='totalPoints_per_price', ascending=False)
      .groupby('position')
      .head(num)
)

# Durchschnittliche Effizienz pro Position aus Top-N
avg_top = top_per_position.groupby('position')['totalPoints_per_price'].mean().reset_index()
avg_top['Typ'] = f'Top {num} per Position'

# Linienpunkte erzeugen
max_price = df['price'].max()
line_points = []

for _, row in avg_top.iterrows():
    eff = row['totalPoints_per_price']
    pos = row['position']
    typ = row['Typ']
    line_points.extend([
        {'price': 0, 'totalPoints': 0, 'position': pos, 'Typ': typ},
        {'price': max_price, 'totalPoints': eff * max_price, 'position': pos, 'Typ': typ},
    ])

line_df = pd.DataFrame(line_points)

# Linienchart (eine Linie pro Position, gestrichelt)
line_chart = alt.Chart(line_df).mark_line(size=2, strokeDash=[5, 3]).encode(
    x='price:Q',
    y='totalPoints:Q',
    color='position:N'
)

# Streudiagramm
scatter = alt.Chart(df).mark_circle(size=60).encode(
    x='price_jitter',
    y='totalPoints',
    color='position',
    tooltip=['displayName', 'position', 'totalPoints', 'averagePoints', 'price']
).properties(
    title=f'Total Points vs Price (Linien: Top {num} Spieler pro Position)',
    width=800,
    height=800
)

# Kombinieren
final_chart = scatter + line_chart
final_chart



alt.LayerChart(...)

In [97]:
chart = alt.Chart(df).mark_circle(size=60).encode(
    x='price',
    y='averagePoints',
    color='position', # Color based on position
    tooltip=['displayName', 'position', 'totalPoints', 'averagePoints', 'price']
).properties(
    title='averagePoints Points vs Price by Position'
)
# Create the diagonal line
line = alt.Chart(pd.DataFrame({'x': [0, 11], 'y': [0, 7.8]}), width=800, height=800).mark_line(
    color='black',
    strokeDash=[2, 2],
    opacity=0.5,
    strokeWidth=1
).encode(
    x='x',
    y='y'
)
# Combine the chart and the line
final_chart = chart + line
# Display the chart
final_chart

alt.LayerChart(...)

In [169]:
import pandas as pd
import altair as alt
import numpy as np

# Top-N Spieler pro Position
num = 10

# Jitter hinzufügen
df['price_jitter'] = df['price'] + np.random.uniform(-0.1, 0.1, size=len(df))
df['averagePoints_per_price'] = df['averagePoints'] / df['price']

# Top-N Spieler pro Position nach averagePoints_per_price
top_per_position = (
    df.sort_values(by='averagePoints_per_price', ascending=False)
      .groupby('position')
      .head(num)
)

# Durchschnittliche Effizienz (averagePoints/price) pro Position
avg_top = top_per_position.groupby('position')['averagePoints_per_price'].mean().reset_index()
avg_top['Typ'] = f'Top {num} per Position'

# Linien vorbereiten (mit Steigung = Effizienz)
max_price = df['price'].max()
line_points = []

for _, row in avg_top.iterrows():
    eff = row['averagePoints_per_price']
    pos = row['position']
    typ = row['Typ']
    line_points.extend([
        {'price': 0, 'averagePoints': 0, 'position': pos, 'Typ': typ},
        {'price': max_price, 'averagePoints': eff * max_price, 'position': pos, 'Typ': typ},
    ])

line_df = pd.DataFrame(line_points)

# Linienchart (gestrichelt)
line_chart = alt.Chart(line_df).mark_line(size=2, strokeDash=[5, 3]).encode(
    x='price:Q',
    y='averagePoints:Q',
    color='position:N'
)

# Streudiagramm (mit Jitter)
scatter = alt.Chart(df).mark_circle(size=60).encode(
    x='price_jitter',
    y='averagePoints',
    color='position',
    tooltip=['displayName', 'position', 'totalPoints', 'averagePoints', 'price', 'minutesPlayed']
).properties(
    title=f'Average Points vs Price (Linien: Top {num} Spieler pro Position)',
    width=800,
    height=800
)

# Kombinieren
final_chart = scatter + line_chart
final_chart


alt.LayerChart(...)

Nun möchten wir das perfekte Team finden. Dafür vereinfachen wir die Constraints etwas:
Da Stürmer im Vergleich zu anderen Positionen deutlich schwächer in Points per Price und Average Points per Price abschneiden, nehmen wir an, dass die optimale Formation entweder ein 5-4-1 oder ein 4-5-1 ist.

Wir versuchen also, die besten 11 Spieler zu finden und die übrigen Positionen mit den günstigsten verfügbaren Spielern aufzufüllen. Im Sturm gibt es nur einen 4-Millionen-Spieler und einige wenige für 4.5 Millionen. Ein schlechter Torwart kostet uns rund 3.5 Millionen. Ergänzt wird die Ersatzbank mit einem beliebigen 4-Millionen-Spieler aus dem Mittelfeld oder der Verteidigung (abhängig von der gewählten Formation).

Daraus ergibt sich:
16 Millionen werden für günstige Ersatzspieler verwendet.
Damit bleiben 84 Millionen für unsere optimale Startelf übrig.

Für diese Optimierung wenden wir den Knapsack-Algorithmus an mit den folgenden Constraints:

    Maximal 5 Verteidiger

    Maximal 5 Mittelfeldspieler

    Insgesamt maximal 9 Spieler aus Verteidigung + Mittelfeld

    Genau 1 Torwart

    Genau 1 Stürmer

    Maximal 3 Spieler pro Team

Ziel: Maximiere die Total Points der Startaufstellung.



In [104]:
!pip install pulp

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 63.4 MB/s eta 0:00:00


In [138]:
import pulp
from collections import Counter

# Problem definieren
problem = pulp.LpProblem("Best_Fantasy_Team", pulp.LpMaximize)

# Binärvariable: 1 = Spieler im Team, 0 = nicht im Team
player_vars = {
    row['displayName']: pulp.LpVariable(row['displayName'], cat='Binary')
    for _, row in df.iterrows()
}

# Zielfunktion: Maximiere totalPoints
problem += pulp.lpSum([
    player_vars[row['displayName']] * row['totalPoints']
    for _, row in df.iterrows()
]), "Total_Team_Points"

# Budget-Beschränkung (max 84 Mio)
problem += pulp.lpSum([
    player_vars[row['displayName']] * row['price']
    for _, row in df.iterrows()
]) <= 84, "Budget"

# Genau 11 Spieler
problem += pulp.lpSum(player_vars.values()) == 11, "Total_Players"

#  Genau 1 Torwart
problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'] == 'GK'].iterrows()
]) == 1, "One_Goalkeeper"

# Genau 1 Stürmer
problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'] == 'FWD'].iterrows()
]) == 1, "One_Forward"

# Max 5 DEF + Max 5 MID + Max 9 DEF+MID
problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'] == 'DEF'].iterrows()
]) <= 5, "Max_5_DEF"

problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'] == 'MID'].iterrows()
]) <= 5, "Max_5_MID"

problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'].isin(['DEF', 'MID'])].iterrows()
]) <= 9, "Max_9_DEF_MID"

# 6️Max 3 Spieler pro Team
for team in df['squadId'].unique():
    problem += pulp.lpSum([
        player_vars[row['displayName']]
        for _, row in df[df['squadId'] == team].iterrows()
    ]) <= 3, f"Max_3_from_{team}"


In [139]:
# Problem lösen
solver = pulp.PULP_CBC_CMD(msg=False)
problem.solve(solver)

# Ergebnis anzeigen
chosen_players = []
for _, row in df.iterrows():
    name = row['displayName']
    if pulp.value(player_vars[name]) == 1:
        chosen_players.append(row)

team_df = pd.DataFrame(chosen_players)
total_cost = team_df['price'].sum()
total_points = team_df['totalPoints'].sum()

print(f"Kosten: {total_cost:.2f} Mio")
print(f"Total Points: {total_points:.0f}")
display(team_df[['displayName', 'position', 'squadId', 'price','averagePoints', 'totalPoints']].sort_values(by='position'))


Kosten: 84.00 Mio
Total Points: 1688


displayName position  squadId  price  averagePoints  totalPoints
9     N. Lavanchy      DEF     1414    5.5            2.8          100
16    N. Dussenne      DEF      490    7.5            4.3          143
70    A. Ciganiks      DEF     1269    5.0            2.6           90
80       M. Poaty      DEF      490    6.0            3.2          117
90      D. Schmid      DEF      481    8.0            4.7          165
176     B. Traoré      FWD      481    9.0            4.7          170
14     Y. Brecher       GK      595    7.5            4.8          181
5      X. Shaqiri      MID      481   11.0            7.8          265
147  M. Di Giusto      MID     3573    8.5            3.9          149
174    A. Sanches      MID      490    8.5            5.6          158
178   D. Rrudhani      MID     1269    7.5            4.8          150

| 	displayName |	position |	squadId |	price |	averagePoints |	totalPoints |
|-|-|-|-|-|-|
|N. Lavanchy 	| DEF 	| 1414 	| 5.5 	|2.8 	|100
|N. Dussenne 	| DEF 	| 490 	| 7.5 	|4.3 |	143
|A. Ciganiks 	| DEF 	| 1269 	| 5.0 	|2.6 	|90
|M. Poaty 	  | DEF 	| 490 	| 6.0 	|3.2 |	117
|	D. Schmid 	| DEF 	| 481 	|8.0 	|4.7 |	165
|B. Traoré  	| FWD 	| 481 	|9.0 	|4.7 	|170
|Y. Brecher 	| GK 	  | 595 	|7.5 	|4.8 	|181
|	X. Shaqiri 	| MID 	| 481 	|11.0 	|7.8 |	265
|	M. Di Giusto 	| MID | 3573 	|8.5 	|3.9 	|149
|	A. Sanches 	| MID 	| 490 	|8.5 |	5.6| 	158
|	D. Rrudhani |	MID 	| 1269 	|7.5 	|4.8 	|150


Kosten: 84.00 Mio
Total Points: 1688

Kosten: 88.00 Mio
Total Points: 1738

    	displayName 	position 	squadId 	price 	averagePoints 	totalPoints
    N. Lavanchy 	  DEF 	1414 	5.5 	2.8 	100
    N. Dussenne 	  DEF 	490 	7.5 	4.3 	143
    A. Ciganiks 	  DEF 	1269 	5.0 	2.6 	90
    M. Poaty 	      DEF 	490 	6.0 	3.2 	117
    D. Schmid 	    DEF 	481 	8.0 	4.7 	165
    B. Traoré 	    FWD 	481 	9.0 	4.7 	170
    T. Berdayes 	  FWD 	1414 	6.5 	2.9 	106
    Y. Brecher 	    GK 	595 	7.5 	4.8 	181
    X. Shaqiri  	  MID 	481 	11.0 	7.8 	265
    B. Toma 	      MID 	453 	6.0 	2.7 	93
    A. Sanches 	    MID 	490 	8.5 	5.6 	158
    D. Rrudhani 	  MID 	1269 	7.5 	4.8 	150

In [171]:
df = df[ (df['minutesPlayed'] > 900)]

In [174]:

# Problem definieren
problem = pulp.LpProblem("Best_Fantasy_Team_by_AvgPoints", pulp.LpMaximize)

# Binärvariable: 1 = Spieler im Team, 0 = nicht im Team
player_vars = {
    row['displayName']: pulp.LpVariable(row['displayName'], cat='Binary')
    for _, row in df.iterrows()
}

# 🎯 Neue Zielfunktion: Maximiere averagePoints
problem += pulp.lpSum([
    player_vars[row['displayName']] * row['averagePoints']
    for _, row in df.iterrows()
]), "Total_AveragePoints"

# Budget-Beschränkung (max 84 Mio)
problem += pulp.lpSum([
    player_vars[row['displayName']] * row['price']
    for _, row in df.iterrows()
]) <= 84, "Budget"

# Genau 11 Spieler
problem += pulp.lpSum(player_vars.values()) == 11, "Total_Players"

# Genau 1 Torwart
problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'] == 'GK'].iterrows()
]) == 1, "One_Goalkeeper"

# Genau 1 Stürmer
problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'] == 'FWD'].iterrows()
]) == 1, "One_Forward"

# Max 5 DEF + Max 5 MID + Max 9 DEF+MID
problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'] == 'DEF'].iterrows()
]) <= 5, "Max_5_DEF"

problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'] == 'MID'].iterrows()
]) <= 5, "Max_5_MID"

problem += pulp.lpSum([
    player_vars[row['displayName']]
    for _, row in df[df['position'].isin(['DEF', 'MID'])].iterrows()
]) <= 9, "Max_9_DEF_MID"

# Max 3 Spieler pro Team
for team in df['squadId'].unique():
    problem += pulp.lpSum([
        player_vars[row['displayName']]
        for _, row in df[df['squadId'] == team].iterrows()
    ]) <= 3, f"Max_3_from_{team}"

# Problem lösen
solver = pulp.PULP_CBC_CMD(msg=False)
problem.solve(solver)

# Ergebnis anzeigen
chosen_players = []
for _, row in df.iterrows():
    name = row['displayName']
    if pulp.value(player_vars[name]) == 1:
        chosen_players.append(row)

team_df = pd.DataFrame(chosen_players)
total_cost = team_df['price'].sum()
avg_points_total = team_df['averagePoints'].sum()/11.0

print(f"Kosten: {total_cost:.2f} Mio")
print(f"Total Average Points (avg of 11): {avg_points_total:.2f}")
display(team_df[['displayName', 'position', 'squadId', 'price', 'averagePoints', 'totalPoints', 'minutesPlayed']].sort_values(by='position'))



Kosten: 84.00 Mio
Total Average Points (avg of 11): 4.86


displayName position  squadId  price  averagePoints  totalPoints  \
4       L. Benito      DEF     1963    5.5            3.2           86   
12    S. Rouiller      DEF      210    6.5            4.0          112   
16    N. Dussenne      DEF      490    7.5            4.3          143   
90      D. Schmid      DEF      481    8.0            4.7          165   
162   G. Koutsias      FWD      935    6.0            3.1           61   
20       J. Frick       GK      210    6.0            4.9           64   
5      X. Shaqiri      MID      481   11.0            7.8          265   
72   C. Fassnacht      MID     1963    9.5            5.8          104   
142      P. Otele      MID      481    8.0            5.3           95   
174    A. Sanches      MID      490    8.5            5.6          158   
178   D. Rrudhani      MID     1269    7.5            4.8          150   

     minutesPlayed  
4             2246  
12            2419  
16            2937  
90            2922  
162           1241  
20            1170  
5             2618  
72            1440  
142           1020  
174           2371  
178           2189

Kosten: 84.00 Mio
Total Average Points (avg of 11): 4.97

	displayName 	position 	squadId 	price 	averagePoints 	totalPoints
12 	S. Rouiller 	DEF 	210 	6.5 	4.0 	112
16 	N. Dussenne 	DEF 	490 	7.5 	4.3 	143
40 	K. Rüegg 	DEF 	481 	5.5 	4.7 	33
90 	D. Schmid 	DEF 	481 	8.0 	4.7 	165
34 	C. Bedia 	FWD 	1963 	7.0 	3.6 	62
20 	J. Frick 	GK 	210 	6.0 	4.9 	64
5 	X. Shaqiri 	MID 	481 	11.0 	7.8 	265
28 	B. Kololli 	MID 	1414 	7.0 	4.5 	95
72 	C. Fassnacht 	MID 	1963 	9.5 	5.8 	104
174 	A. Sanches 	MID 	490 	8.5 	5.6 	158
178 	D. Rrudhani 	MID 	1269 	7.5 	4.8 	150


*   List item
*   List item



Kosten: 84.00 Mio
Total Average Points (avg of 11): 4.97

	displayName 	position 	squadId 	price 	averagePoints 	totalPoints
12 	S. Rouiller 	DEF 	210 	6.5 	4.0 	112
16 	N. Dussenne 	DEF 	490 	7.5 	4.3 	143
40 	K. Rüegg 	DEF 	481 	5.5 	4.7 	33
90 	D. Schmid 	DEF 	481 	8.0 	4.7 	165
34 	C. Bedia 	FWD 	1963 	7.0 	3.6 	62
20 	J. Frick 	GK 	210 	6.0 	4.9 	64
5 	X. Shaqiri 	MID 	481 	11.0 	7.8 	265
28 	B. Kololli 	MID 	1414 	7.0 	4.5 	95
72 	C. Fassnacht 	MID 	1963 	9.5 	5.8 	104
174 	A. Sanches 	MID 	490 	8.5 	5.6 	158
178 	D. Rrudhani 	MID 	1269 	7.5 	4.8 	150


In [175]:
my_players = [
    "M. Hitz", #-> brecher
    #"A. Spagnoli",
    "D. Schmid",
    "B. Freimann",
    "N. Dussenne",
    "S. Rouiller",
    "N. Lavanchy",
    "X. Shaqiri",
    "M. Stevanovic",
    "D. Rrudhani",
    "C. Fassnacht",
    #"S. Stroscio",
    "T. Berdayes",
    #"A. Duville-Parsemain",
    #"L. Kabashi"
]

# Filtere Spieler aus dem DataFrame
my_team_df = df[df['displayName'].isin(my_players)].copy()

my_team_df = my_team_df[my_team_df['firstName']!='Mihailo']

# Berechne Kennzahlen
total_cost = my_team_df['price'].sum()
total_points = my_team_df['totalPoints'].sum()
avg_points_sum = my_team_df['averagePoints'].sum()/len(my_team_df)

# Ausgabe
print(f"👥 Anzahl Spieler: {len(my_team_df)}")
print(f"💰 Total Cost: {total_cost:.2f} Mio")
print(f"⭐ Total Points: {total_points:.0f}")
print(f"📊 Sum of Average Points: {avg_points_sum:.2f}")

# Optional: anzeigen
display(my_team_df[['displayName', 'position', 'price', 'totalPoints', 'averagePoints']].sort_values(by='position'))

👥 Anzahl Spieler: 11
💰 Total Cost: 84.00 Mio
⭐ Total Points: 1599
📊 Sum of Average Points: 4.59


displayName position  price  totalPoints  averagePoints
9      N. Lavanchy      DEF    5.5          100            2.8
12     S. Rouiller      DEF    6.5          112            4.0
16     N. Dussenne      DEF    7.5          143            4.3
90       D. Schmid      DEF    8.0          165            4.7
292    B. Freimann      DEF    4.5           47            2.4
177    T. Berdayes      FWD    6.5          106            2.9
2          M. Hitz       GK    7.5          183            5.1
5       X. Shaqiri      MID   11.0          265            7.8
7    M. Stevanovic      MID   10.0          224            5.9
72    C. Fassnacht      MID    9.5          104            5.8
178    D. Rrudhani      MID    7.5          150            4.8